File: erlang/take_drop.erl
%% file: take_drop.erl
-module(take_drop).
-export([take/2, drop/2]).

take([], _) -> [];
take([H|T], N) when N >0 ->
    [H|take(T, N - 1)];
take(_, 0) -> [].

drop([], _) -> [];
drop([_|T], N) when N > 0 ->
    drop(T, N - 1);
drop(Xs, 0) -> Xs.

File: prolog/take-drop.pl
%% file: take_drop.pl
take(N, Xs, Ys) :-
    N >= 0, take(N, Xs, Zs, []), reverse(Zs, Ys).

take(_, [], Acc, Acc).
take(0, _, Acc, Acc).
take(N, [H|T], Ys, Acc) :-
    N >= 1, N1 is N - 1, take(N1, T, Ys, [H|Acc]), !.


drop(0, Xs, Xs).
drop(_, [], []).
drop(N, [_|T], Ys) :-
    N >= 0, N1 is N - 1, drop(N1, T, Ys), !.
    
